File: 2025/q06.jl
#!/usr/bin/env julia

function part1(data)
    ans = 0
    a = 0
    for x in data
        x == 'A' && (a += 1)
        x == 'a' && (ans += a)
    end
    return ans
end


function part2(data)
    ans = 0
    n = Dict{Char,Int}()
    for x in data
        if isuppercase(x)
            n[x] = 1 + get(n, x, 0)
        else
            ans += get(n, uppercase(x), 0)
        end
    end
    return ans
end


function part3(data, N=1000, T=1000)
    data = collect(data ^ N)
    ans = 0
    for (i, x) in enumerate(data)
        if islowercase(x)
            s = max(1, i-T)
            e = min(length(data), i+T)
            ans += count(==(uppercase(x)), data[s:e])
        end
    end
    return ans
end


function part3(data, N=1000, T=1000)
    m = length(data)
    e = m * N
    n = Dict{Char,Int}()
    for x in data[1:T]
        isuppercase(x) && (n[x] = 1 + get(n, x, 0))
    end
    ans = 0
    for i in 1:e
        k, j = i-T-1, i+T
        if k >= 1
            x = data[mod(k, 1:m)]
            isuppercase(x) && (n[x] -= 1)
        end
        if j <= e
            x = data[mod(j, 1:m)]
            isuppercase(x) && (n[x] = 1 + get(n, x, 0))
        end
        x = data[mod(i, 1:m)]
        if islowercase(x)
            ans += get(n, uppercase(x), 0)
        end
    end
    return ans
end


@assert part1("ABabACacBCbca") == 5
@assert part2("ABabACacBCbca") == 11
@assert part3("AABCBABCABCabcabcABCCBAACBCa", 1, 10) == 34
@assert part3("AABCBABCABCabcabcABCCBAACBCa", 2, 10) == 72


data = readchomp("q06_p1.txt")
println("part1: ", part1(data))
data = readchomp("q06_p2.txt")
println("part2: ", part2(data))
data = readchomp("q06_p3.txt")
println("part3: ", @time part3(data))

File: event/2024/q06.py
#!/usr/bin/env python
from collections import Counter


def part1(data):
    par = dict()
    fru = list()
    for line in data.strip().splitlines():
        n, ps = line.split(':')
        ps = ps.split(',')
        for p in ps:
            if p == '@':
                fru.append(n)
            else:
                par[p] = n
    paths = list()
    stats = Counter()
    for s in fru:
        path = [s]
        while s != 'RR':
            s = par[s]
            path.append(s)
        paths.append(path)
        stats[len(path)] += 1
    n,_ = stats.most_common()[-1]
    best = next(p for p in paths if len(p) == n)
    ans = ''.join(best[::-1]) + '@'
    return ans


def part2(data):
    par = dict()
    fru = list()
    for line in data.strip().splitlines():
        n, ps = line.split(':')
        ps = ps.split(',')
        for p in ps:
            if p == '@':
                fru.append(n)
            else:
                par[p] = n
    paths = dict()
    stats = Counter()
    for f in fru:
        path = 1
        s = f
        while s != 'RR':
            s = par[s]
            path += 1
        stats[path] += 1
        paths[f] = path
    n,_ = stats.most_common()[-1]
    s = next(f for f,p in paths.items() if p == n)
    ans = s[0] + '@'
    while s != 'RR':
        s = par[s]
        ans = s[0] + ans
    return ans


def part3(data):
    par = dict()
    fru = list()
    for line in data.strip().splitlines():
        n, ps = line.split(':')
        if n in ('BUG', 'ANT'): continue
        ps = ps.split(',')
        for p in ps:
            if p == '@':
                fru.append(n)
            elif p not in ('BUG', 'ANT'):
                par[p] = n
    paths = dict()
    stats = Counter()
    for f in fru:
        path = 1
        s = f
        while s != 'RR':
            s = par[s]
            path += 1
        stats[path] += 1
        paths[f] = path
    n,_ = stats.most_common()[-1]
    s = next(f for f,p in paths.items() if p == n)
    ans = s[0] + '@'
    while s != 'RR':
        s = par[s]
        ans = s[0] + ans
    return ans


data = """
RR:A,B,C
A:D,E
B:F,@
C:G,H
D:@
E:@
F:@
G:@
H:@
"""
assert part1(data) == 'RRB@'
assert part2(data) == 'RB@'
assert part3(data) == 'RB@'


data = open('q06_p1.txt').read()
ans = part1(data)
print('part1:', ans)

data = open('q06_p2.txt').read()
ans = part2(data)
print('part2:', ans)

data = open('q06_p3.txt').read()
ans = part3(data)
print('part3:', ans)
